core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala

```diff
@@ -70,23 +72,174 @@ object H2OSchemaUtils {
     StructType(types)
   }
 
-  def flattenSchema(schema: StructType, prefix: String = null, nullable: Boolean = false): StructType = {
+  def flattenDataFrame(df: DataFrame): DataFrame = {
+    val schema = flattenSchema(df)
+    flattenDataFrame(df, schema)
+  }
+
+  def flattenDataFrame(df: DataFrame, flatSchema: StructType): DataFrame = {
+    implicit val rowEncoder = RowEncoder(flatSchema)
+    val numberOfColumns = flatSchema.fields.length
+    val flatSchemaIndexes = flatSchema.fields.map(_.name).zipWithIndex.toMap
+    val originalSchema = df.schema
+    df.map[Row]{ r: Row =>
+      val result = ArrayBuffer.fill[Any](numberOfColumns)(null)
+      val fillMethod = fillBuffer(flatSchemaIndexes, result, r)(_, _, _)
+      originalSchema.fields.zipWithIndex.foreach{ case (f, i) => fillMethod(f, i, None)}
+      Row.fromSeq(result)
+    }
+  }
+
+  private def fillBuffer
+      (flatSchemaIndexes: Map[String, Int], buffer: ArrayBuffer[Any], row: Row)
+      (field: StructField, index: Int, prefix: Option[String] = None): Unit = {
+    val StructField(name, dataType, _, _) = field
+    val qualifiedName = getQualifiedName(prefix, name)
+    if(!row.isNullAt(index)) {
+      dataType match {
+        case BinaryType =>
+          fillArray(flatSchemaIndexes, buffer, row, ByteType, index, qualifiedName)
+        case MapType(_, valueType, containsNull) =>
+          val map = row.getMap[Any, Any](index)
+          val subRow = Row.fromSeq(map.values.toSeq)
+          val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+          map.keys.zipWithIndex.foreach{ case (k, i) =>
+            val mapField = StructField(k.toString, valueType, containsNull)
+            fillMethod(mapField, i, Some(qualifiedName))
+          }
+        case ArrayType(elementType, _) =>
+          fillArray(flatSchemaIndexes, buffer, row, elementType, index, qualifiedName)
+        case StructType(fields) =>
+          val subRow = row.getStruct(index)
+          val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+          fields.zipWithIndex.foreach { case (f, i) => fillMethod(f, i, Some(qualifiedName)) }
+        case dt => buffer(flatSchemaIndexes(qualifiedName)) = row.get(index)
+      }
+    }
+  }
+
+  private def fillArray(
+      flatSchemaIndexes: Map[String, Int],
+      buffer: ArrayBuffer[Any],
+      row: Row,
+      elementType: DataType,
+      index: Int,
+      qualifiedName: String){
+    val seq = row.getSeq[Any](index)
+    val subRow = Row.fromSeq(seq)
+    val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+    (0 until seq.size).foreach{ i =>
+      val arrayField = StructField(i.toString, elementType)
+      fillMethod(arrayField, i, Some(qualifiedName))
+    }
+  }
+
+  def flattenSchema(df: DataFrame): StructType = {
+    val originalSchema = df.schema
+    val fields = df.rdd.map[Seq[FieldWithOrder]] { row: Row =>
```

Comment:
I was thinking -> why do we need to process schema on per-row basis? Can't we just process schema from `df.schema` directly?

Reply:
it's because of variable length of arrays and also arrays might be arbitrarily nested. In this solution the `map` calculates a specific schema for each record and `reduce` does the merge logic.

Reply:
thank you for the explanation!